core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -232,6 +233,16 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
         if (canExecuteScript(lang, scriptEngineService, script.getType(), scriptContext) == false) {
             throw new ScriptException("scripts of type [" + script.getType() + "], operation [" + scriptContext.getKey() + "] and lang [" + lang + "] are disabled");
         }
+
+        // special exception to prevent expressions from compiling as update or mapping scripts
+        boolean expression = scriptEngineService instanceof ExpressionScriptEngineService;
```

Comment:
Sure, I would agree with pushing it out to a different PR, but I don't understand why its a big or a breaking change. The only current ScriptEngineService that need to return anything special is the ExpressionScriptEngineService since it would need to return all ScriptContext values except UPDATE and MAPPING. All other implementations could return all values of ScriptContext. This could probably even be a public static constant so they can all return the same instance.